packages/next/server/web/spec-extension/response.ts

```diff
@@ -44,23 +44,40 @@ export class NextResponse extends Response {
     })
   }
 
-  static redirect(url: string | NextURL | URL, status = 307) {
+  static redirect(
+    url: string | NextURL | URL,
+    status = 307,
+    init: ResponseInit = {}
+  ) {
     if (!REDIRECTS.has(status)) {
       throw new RangeError(
         'Failed to execute "redirect" on "response": Invalid status code'
       )
     }
 
     return new NextResponse(null, {
-      headers: { Location: validateURL(url) },
+      ...init,
+      headers: {
+        ...init.headers,
+        Location: validateURL(url),
+      },
       status,
     })
   }
 
-  static rewrite(destination: string | NextURL | URL) {
-    return new NextResponse(null, {
-      headers: { 'x-middleware-rewrite': validateURL(destination) },
-    })
+  static rewrite(destination: string | NextURL | URL, init: ResponseInit = {}) {
+    if (init.headers) {
+      init.headers = {
+        ...init.headers,
+        'x-middleware-rewrite': validateURL(destination),
+      }
+    } else {
+      init.headers = {
+        'x-middleware-rewrite': validateURL(destination),
+      }
+    }
+
+    return new NextResponse(null, init)
```

Comment:
```suggestion
    return new NextResponse(null, {
      ...init,
      headers: {
        ...init.headers,
        'x-middleware-rewrite': validateURL(destination)
      },
    })
```

Reply:
What if init parameter does not contain headers property?

Reply:
> What if init parameter does not contain headers property?

@FLCN-16

I submitted the review on GitHub.com website so my review might lack a bit of accuaracy.

It should be `...init?.headers`.